include/ghex/structured/field_descriptor.hpp

```diff
@@ -0,0 +1,254 @@
+          * @return reference to the value in the buffer */
+        GT_FUNCTION
+        T& buffer(const coordinate_type& coord) const noexcept {
+            // compute buffer coordinates, relative to the buffer origin
+            const coordinate_type buffer_coord = coord - m_data_is.m_first;
+            // dot product with strides to compute address
+            const auto memory_location = dot(m_buffer_desc.m_strides, buffer_coord);
+            return *reinterpret_cast<T*>(
+                reinterpret_cast<char*>(m_buffer_desc.m_ptr) + memory_location);
+        }
+
+        /** @brief accesses field at specified local coordinate
+          * @param coord in local coordinate system
+          * @return const reference to the value in the field */
+        GT_FUNCTION
+        const T& data(const coordinate_type& coord) const noexcept {
+            // make data memory coordinates from local coordinates
+            const coordinate_type data_coord = coord + m_data_is.m_offset;
+            // dot product with strides to compute address
+            const auto memory_location = dot(m_data_is.m_strides, data_coord);
+            return *reinterpret_cast<const T*>(
+                reinterpret_cast<const char*>(m_data_is.m_ptr) + memory_location);
+        }
+
+    };
+
+    struct unpack_iteration_space {
+        using value_t = T;
+        using coordinate_t = coordinate_type;
+        const buffer_descriptor<const T*> m_buffer_desc;
+        const basic_iteration_space<T*> m_data_is;
+
+        /** @brief accesses buffer at specified local coordinate
+          * @param coord in local coordinate system
+          * @return value in the buffer */
+        GT_FUNCTION
+        const T& buffer(const coordinate_type& coord) const noexcept {
+            // compute buffer coordinates, relative to the buffer origin
+            const coordinate_type buffer_coord = coord - m_data_is.m_first;
+            // dot product with strides to compute address
+            const auto memory_location = dot(m_buffer_desc.m_strides, buffer_coord);
+            return *reinterpret_cast<const T*>(
+                reinterpret_cast<const char*>(m_buffer_desc.m_ptr) + memory_location);
+        }
+
+        /** @brief accesses field at specified local coordinate
+          * @param local coordinate system
+          * @return reference to the value in the field */
+        GT_FUNCTION
+        T& data(const coordinate_type& coord) const noexcept {
+            // make data memory coordinates from local coordinates
+            const coordinate_type data_coord = coord + m_data_is.m_offset;
+            // dot product with strides to compute address
+            const auto memory_location = dot(m_data_is.m_strides, data_coord);
+            return *reinterpret_cast<T*>(
+                reinterpret_cast<char*>(m_data_is.m_ptr) + memory_location);
+        }
+    };
+
+protected: // members
+    domain_descriptor_type m_dom;             ///< domain descriptor
```

Comment:
good question. I don't see a problem with a copy here (it's probably even safer from a dangling reference point) since the domain descriptor does not allocate and is cheap to copy.

Reply:
Ok, sure.